### PR TITLE
[SPARK-15079] Support average/count/sum in Long/DoubleAccumulator

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?

This patch removes AverageAccumulator and adds the ability to compute average to LongAccumulator and DoubleAccumulator. The patch also improves documentation for the two accumulators.
## How was this patch tested?

Added unit tests for this.

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan

File: core/src/main/scala/org/apache/spark/AccumulatorV2.scala [hunk 1]
```diff
@@ -257,23 +257,66 @@ private[spark] object AccumulatorContext {
...
+ * An [[AccumulatorV2 accumulator]] for computing sum, count, and averages for 64-bit integers.
+ *
+ * @since 2.0.0
+ */
 class LongAccumulator extends AccumulatorV2[jl.Long, jl.Long] {
   private[this] var _sum = 0L
+  private[this] var _count = 0L
 
-  override def isZero: Boolean = _sum == 0
+  /**
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/AccumulatorV2.scala]: wrong doc for `isZero`

File: core/src/main/scala/org/apache/spark/AccumulatorV2.scala [hunk 2]
```diff
@@ -257,23 +257,66 @@ private[spark] object AccumulatorContext {
...
+   * Returns the sum of elements added to the accumulator.
+   * @since 2.0.0
+   */
   def sum: Long = _sum
 
+  /**
+   * Returns the average of elements added to the accumulator.
+   * @since 2.0.0
+   */
+  def avg: Double = _sum.toDouble / _count
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/AccumulatorV2.scala]: should we handle `_count == 0` here?

**Comment by cloud-fan** [comment]
LGTM

File: core/src/main/scala/org/apache/spark/AccumulatorV2.scala (hunk 2 above)
Comment (rxin) [inline core/src/main/scala/org/apache/spark/AccumulatorV2.scala]: it just becomes Infinity or NaN.

**Comment by rxin** [comment]
Merging in master / branch-2.0.

**Comment by HyukjinKwon** [comment]
Hi @cloud-fan , Could I please ask a question? It seems old accumulator is deprecated but new accumulator should be used everywhere. I am trying to change some tests and codes to use this but I am a bit confused if I should use `LongAccumulator` for `Accumulator[Int]`.

I thought I should leave `Accumulator[Int]` as it is but it seems some tests in `AccumulatorV2Suite` uses int (eg. `0` not `0L`). 

So.. Should I use `LongAccumulator` for `Accumulator[Int]` or just leave them? It seems there is no implementation of `IntAccumulator`. This is intendedly not implemented?

**Comment by cloud-fan** [comment]
Conceptually long accumulator is enough for integral accumulation, that's why we don't implement `IntAccumulator`. However, `LongAccumulator` can't be used as `Accumulator[Int]`, the type system doesn't allow it.

Some tests in `AccumulatorV2Suite` uses int (eg. 0 not 0L) because compiler can cast it to long for us. (that why we think int accumulator is not that useful, as int can always be casted to long)

**Comment by HyukjinKwon** [comment]
@cloud-fan Ah, thank you so much for a detailed explanation. I will look into this for myself.

**Pull request closed without merging**